File: hotel.py
class HotelReservation:
    
    global hotels 
    
    hotels = {
        'green_valley' : {
            'rating' : 3,
            'regular' : {
                'weekday' : 1100,
                'weekend' : 900,
            },
            'rewards' : {
                'weekday' : 800,
                'weekend' : 800,
            }
        },
        'red_river' : {
            'rating' : 4,
            'regular' : {
                'weekday' : 1600,
                'weekend' : 600,
            },
            'rewards' : {
                'weekday' : 1100,
                'weekend' : 500,
            }
        },
        'blue_hills' : {
            'rating' : 5,
            'regular' : {
                'weekday' : 2200,
                'weekend' : 1500,
            },
            'rewards' : {
                'weekday' : 1000,
                'weekend' : 400,
            }
        }
    }

        
    def calculate_price(self,input):
    
        customer_type = input.split(": ")[0].lower().strip()
        later_date_part = input.split(": ")[1]
        individual_date = list(map(str, later_date_part.split(", ")))
    
        final_price_arr = {}
        green_valley_price = 0
        red_river_price = 0
        blue_hills_price = 0
    
        for i in individual_date:
            check_day = i.split("(")[1][:-1]
            if check_day == 'sat' or check_day == 'sun':
                day = 'weekend'
            else:
                day = 'weekday'
            
            green_valley_price += hotels['green_valley'][customer_type][day]
            red_river_price += hotels['red_river'][customer_type][day]
            blue_hills_price += hotels['blue_hills'][customer_type][day]
        
        final_price_arr['green_valley'] = green_valley_price
        final_price_arr['red_river'] = red_river_price
        final_price_arr['blue_hills'] = blue_hills_price
        
        min_value = min(final_price_arr.values())
        min_keys = [k for k in final_price_arr if final_price_arr[k] == min_value]
        
        max_value = {}
        for i in min_keys:
            max_value[i] = hotels[i]['rating']
            
        return max(max_value, key=max_value.get)
        

obj = HotelReservation()
cheapest_hotel = obj.calculate_price(input())

print("The cheapest hotel is: " + cheapest_hotel)

File: hotel.php
<?php 

class HotelReservation {


	var $weekends = ["sat", "sun"];

	var $hotel = [

		'GreenValley' => [

				'rating' => 3,
				'weekdays' => [
					'regular' => 1100,
					'rewards' => 800,
				],
				'weekends' => [
					'regular' => 900,
					'rewards' => 800,
				],
		],

		'RedRiver' => [

				'rating' => 4,
				'weekdays' => [
					'regular' => 1600,
					'rewards' => 1100,
				],
				'weekends' => [
					'regular' => 600,
					'rewards' => 500,
				],
		],

		'BlueHills' => [

				'rating' => 5,
				'weekdays' => [
					'regular' => 2200,
					'rewards' => 1000,
				],
				'weekends' => [
					'regular' => 1500,
					'rewards' => 400,
				],
		],

	];


	public function cheapestHotel($inputs) {

		$inputs = trim($inputs);

		$cust_type = strtolower(explode(":", $inputs)[0]);
		$range_of_dates = explode(":", $inputs)[1];
		
		$dates = explode("," ,$range_of_dates);

		$sumHArray = array();
		foreach ($this->hotel as $hKey => $HValue) {
			$sumHotel = 0;
			foreach( $dates as $d ) {
				$checkFlag = false;
				foreach( $this->weekends as $w ) {
					if(stripos($d, $w) !== false) {
						$checkFlag = $w;
					}
				}
				if ($checkFlag !=false) {
					$sumHotel += $HValue['weekends'][$cust_type];
				}else{
					$sumHotel += $HValue['weekdays'][$cust_type];
				}
			}
			$sumHArray[$hKey] = $sumHotel;
		}

		$minAmount = min($sumHArray);
		

		/*

		Method 1 to compare similar values if prices match
		$array_comp = [];
		foreach ($sumHArray as $sumHKey => $sumHKval) {
			
			if($sumHKval == $minAmount) {
				$array_comp[$sumHKey] = $this->hotel[$sumHKey]['rating'];
			}

		}

		var_dump(max($array_comp));
		var_dump(array_search(max($array_comp), $array_comp));
		*/


		//Method 2 to compare similar values if prices match
		$array_comp = array_keys($sumHArray, $minAmount);

		$result = array();
		foreach ($array_comp as $key => $value) {
			
			$result[$value] = $this->hotel[$value]['rating'];

		}

		return array_search(max($result), $result);

		
	}


}


$obj = new HotelReservation();

$result = $obj->cheapestHotel("Rewards: 26Mar2009(thur), 27Mar2009(fri),
28Mar2009(sat)");

echo "The cheapest hotel you could find is " . $result;

